src/Simulation/Common/IQuantumProcessor.cs

```diff
@@ -391,6 +391,41 @@ public interface IQuantumProcessor
         /// <param name="qubits">Register to apply the exponent to.</param>
         void ControlledExpFrac(IQArray<Qubit> controls, IQArray<Pauli> paulis, long numerator, long power, IQArray<Qubit> qubits);
 
+        /// <summary>
+        /// Used as extensibility mechanism to add custom operations to IQuantumProcessor. 
+        /// </summary>
+        /// <remarks>
+        /// The custom operation may use only some of the parameters. Depending on the nature of the operation, all of them may be optional. 
+        /// A variety of parameter types are provided for flexibility.
+        /// </remarks>
+        /// <param name="operationId">The id of the custom operation. It is up to the custom operation implementor to define the meaning.</param>
+        /// <param name="qubits">Array of qubits to apply the operation to. Their meaning is specific to the custom operation.</param>
+        /// <param name="paulis">Array of single-qubit Pauli values. Their meaning is specific to the custom operation.</param>
+        /// <param name="longs">Array of long parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="doubles">Array of double parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="bools">Array of bool parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="results">Array of Result parameters. Their meaning is specific to the custom operation.</param>
+        /// <param name="strings">Array of string parameters. Their meaning is specific to the custom operation.</param>
+        void CustomOperation(long operationId, IQArray<Qubit> qubits, IQArray<Pauli> paulis, IQArray<long> longs, IQArray<double> doubles, IQArray<bool> bools, IQArray<Result> results, IQArray<string> strings);
```

Comment:
>CustomOperation [](start = 13, length = 15)

What are the first use cases for this?

Reply:
The immediate use cases are for table lookup for state preparation which is used in benchmarks for one of our projects, but will be used in any state preparation also. Another current use case is for testing reversible circuits for arithmetic functions, which are, for example, used in the same benchmarks.

---
In reply to: [413474648](https://github.com/microsoft/qsharp-runtime/pull/172#discussion_r413474648) [](ancestors = 413474648)

Reply:
To make it more concrete, could you please give the Q# signature for these cases, and why these operations should be intrinsic for benchmarking opposed to being decompositions?